internal/worker/events.go

```diff
@@ -0,0 +1,172 @@
+package worker
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"go.sia.tech/renterd/api"
+	"go.sia.tech/renterd/webhooks"
+	"go.uber.org/zap"
+)
+
+type (
+	EventManager interface {
+		AddSubscriber(id string, s EventSubscriber) (chan struct{}, error)
+		HandleEvent(event webhooks.Event)
+		Run(ctx context.Context, eventURL string, opts ...webhooks.HeaderOption) error
+		Shutdown(context.Context) error
+	}
+
+	EventSubscriber interface {
+		HandleEvent(event webhooks.Event) error
+		Subscribe(e EventManager) error
+	}
+
+	WebhookManager interface {
+		RegisterWebhook(ctx context.Context, wh webhooks.Webhook) error
+		UnregisterWebhook(ctx context.Context, wh webhooks.Webhook) error
+	}
+)
+
+type (
+	eventManager struct {
+		webhooks WebhookManager
+		logger   *zap.SugaredLogger
+
+		registerInterval time.Duration
+
+		mu         sync.Mutex
+		subs       map[string]eventSubscriber
+		registered []webhooks.Webhook
+	}
+
+	eventSubscriber struct {
+		EventSubscriber
+		readyChan chan struct{}
+	}
+)
+
+func NewEventManager(w WebhookManager, l *zap.Logger, registerInterval time.Duration) EventManager {
+	return &eventManager{
+		webhooks: w,
+		logger:   l.Sugar().Named("events"),
+
+		registerInterval: registerInterval,
+
+		subs: make(map[string]eventSubscriber),
+	}
+}
+
+func (e *eventManager) AddSubscriber(id string, s EventSubscriber) (chan struct{}, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	_, ok := e.subs[id]
+	if ok {
+		return nil, fmt.Errorf("subscriber with id %v already exists", id)
+	}
+
+	readyChan := make(chan struct{})
+	if len(e.registered) > 0 {
+		close(readyChan)
+	}
+
+	e.subs[id] = eventSubscriber{s, readyChan}
+	return readyChan, nil
+}
+
+func (e *eventManager) HandleEvent(event webhooks.Event) {
```

Comment:
(nit)) could prepare the logger with the event and module fields up here to save a few lines in the `if/else`.